File: general-server/src/main/java/com/song/middleware/server/web/api/ServiceInfoCollectorApi.java
package com.song.middleware.server.web.api;

import com.song.middleware.server.domain.ServiceDO;
import com.song.middleware.server.storage.ServiceStorageComponent;
import com.song.middleware.server.storage.exception.StorageException;
import com.song.middleware.server.web.support.ApiResult;
import java.util.Collections;
import java.util.List;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

/**
 * Created by song on 2017/8/5.
 */
@Slf4j
@RestController
@RequestMapping(value = "/api/v1/service")
public class ServiceInfoCollectorApi {

    private static final String DEFAULT_SERVICE_NAME = "unknown";
    private final ServiceStorageComponent storageComponent;

    @Autowired
    public ServiceInfoCollectorApi(ServiceStorageComponent storageComponent) {
        this.storageComponent = storageComponent;
    }

    @PostMapping(value = "/query")
    public ApiResult<List<ServiceDO>> query(
        @RequestParam(value = "name", required = false) String name) {
        ApiResult<List<ServiceDO>> apiResult = ApiResult.create();
        if (StringUtils.isEmpty(name)) {
            apiResult.setData(Collections.emptyList());
        }
        try {
            List<ServiceDO> serviceDOS = storageComponent.syncQuery(name);
            if (serviceDOS != null) {
                apiResult.setSuccess(true);
                apiResult.setData(serviceDOS);
            }
        } catch (Exception e) {
            apiResult.setMessage("query failed, " + e.getMessage());
            log.error("query failed_" + e.getMessage(), e);
        }
        return apiResult;
    }

    @PostMapping(value = "/register")
    public ApiResult<ServiceDO> register(
        @RequestParam(value = "name", required = false) String name,
        @RequestParam(value = "description", required = false) String description,
        @RequestParam(value = "ip", required = false) String ip,
        @RequestParam(value = "port", required = false) String port,
        @RequestParam(value = "healthCheckUrl", required = false) String healthCheckUrl) {
        ApiResult<ServiceDO> apiResult = ApiResult.create();
        if (StringUtils.isEmpty(name)) {
            name = DEFAULT_SERVICE_NAME;
        }
        if (StringUtils.isEmpty(ip) || StringUtils.isEmpty(port)) {
            apiResult.setMessage("Either ip or port is empty,please check!");
            return apiResult;
        }
        ServiceDO serviceDO = new ServiceDO();
        serviceDO.setName(name);
        serviceDO.setDescription(description);
        serviceDO.setIp(ip);
        serviceDO.setPort(port);
        serviceDO.setHealthCheckUrl(healthCheckUrl);
        try {
            storageComponent.syncRegister(serviceDO);
            apiResult.setSuccess(true);
        } catch (Exception e) {
            apiResult.setMessage(
                "save into storage failed ,please check the server log,detail:" + e.getMessage());
            log.error("Save service info failed_" + e.getMessage(), e);
        }
        return apiResult;
    }


    @PostMapping(value = "/unregister")
    public ApiResult<Void> unregister(
        @RequestParam(value = "name", required = false) String name,
        @RequestParam(value = "ip", required = false) String ip,
        @RequestParam(value = "port", required = false) String port) {
        ApiResult<Void> apiResult = ApiResult.create();
        if (StringUtils.isNotBlank(name)
            || StringUtils.isBlank(ip)
            || StringUtils.isBlank(port)) {
            apiResult.setMessage("Either name、ip or port is empty,please check.");
            return apiResult;
        }
        try {
            storageComponent.syncUnregister(name, ip, port);
            apiResult.setSuccess(true);
        } catch (StorageException e) {
            log.error(name + "/" + ip + ":" + port + " unregister failed_" + e.getMessage(), e);
        }
        return apiResult;
    }

}

File: general-server/src/main/java/com/song/middleware/server/web/api/ServiceInfoCollectorApi.kt
package com.song.middleware.server.web.api

import com.song.middleware.server.domain.ServiceDO
import com.song.middleware.server.storage.ServiceStorageComponent
import com.song.middleware.server.storage.exception.StorageException
import com.song.middleware.server.web.support.ApiResult
import org.apache.commons.lang3.StringUtils
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

/**
 * Created by song on 2017/8/5.
 */
@RestController
@RequestMapping(value = "/api/v1/service")
class ServiceInfoCollectorApi @Autowired
constructor(private val storageComponent: ServiceStorageComponent) {

    @PostMapping(value = "/query")
    fun query(
            @RequestParam(value = "name", required = false) name: String): ApiResult<List<ServiceDO>> {
        val apiResult = ApiResult.create<List<ServiceDO>>()
        if (StringUtils.isEmpty(name)) {
            apiResult.data = emptyList<ServiceDO>()
        }
        try {
            val serviceDOS = storageComponent.syncQuery(name)
            if (serviceDOS != null) {
                apiResult.isSuccess = true
                apiResult.data = serviceDOS
            }
        } catch (e: Exception) {
            apiResult.message = "query failed, " + e.message
            log.error("query failed_" + e.message, e)
        }

        return apiResult
    }

    @PostMapping(value = "/register")
    fun register(
            @RequestParam(value = "name", required = false) name: String,
            @RequestParam(value = "description", required = false) description: String,
            @RequestParam(value = "ip", required = false) ip: String,
            @RequestParam(value = "port", required = false) port: String,
            @RequestParam(value = "healthCheckUrl", required = false) healthCheckUrl: String): ApiResult<ServiceDO> {
        var name = name
        val apiResult = ApiResult.create<ServiceDO>()
        if (StringUtils.isEmpty(name)) {
            name = DEFAULT_SERVICE_NAME
        }
        if (StringUtils.isEmpty(ip) || StringUtils.isEmpty(port)) {
            apiResult.message = "Either ip or port is empty,please check!"
            return apiResult
        }
        val serviceDO = ServiceDO()
        serviceDO.name = name
        serviceDO.description = description
        serviceDO.ip = ip
        serviceDO.port = port
        serviceDO.healthCheckUrl = healthCheckUrl
        try {
            storageComponent.syncRegister(serviceDO)
            apiResult.isSuccess = true
        } catch (e: Exception) {
            apiResult.message = "save into storage failed ,please check the server log,detail:" + e.message
            log.error("Save service info failed_" + e.message, e)
        }

        return apiResult
    }


    @PostMapping(value = "/unregister")
    fun unregister(
            @RequestParam(value = "name", required = false) name: String,
            @RequestParam(value = "ip", required = false) ip: String,
            @RequestParam(value = "port", required = false) port: String): ApiResult<Void> {
        val apiResult = ApiResult.create<Void>()
        if (StringUtils.isNotBlank(name)
                || StringUtils.isBlank(ip)
                || StringUtils.isBlank(port)) {
            apiResult.message = "Either name、ip or port is empty,please check."
            return apiResult
        }
        try {
            storageComponent.syncUnregister(name, ip, port)
            apiResult.isSuccess = true
        } catch (e: StorageException) {
            log.error(name + "/" + ip + ":" + port + " unregister failed_" + e.message, e)
        }

        return apiResult
    }

    companion object {

        private val log = LoggerFactory.getLogger(ServiceInfoCollectorApi.javaClass)

        private val DEFAULT_SERVICE_NAME = "unknown"
    }

}